### PR TITLE
Converted ZC_ENTER_ROOM to struct

### PR DESCRIPTION
**Pull request by AoShinRO**
* **Addressed Issue(s)**: 

* **Server Mode**: both

* **Description of Pull Request**: Converted ZC_ENTER_ROOM to packet

### CODE REVIEW

File: src/map/clif.cpp [hunk 1]
```diff
@@ -4557,38 +4557,39 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
-			WFIFOL(fd, 8+i*28) = (i != 0 || cd->owner->type == BL_NPC);
-			safestrncpy(WFIFOCP(fd, 8+(i+t)*28+4), cd->usersd[i]->status.name, NAME_LENGTH);
+		for (int i = 0; i < cd.users; i++) {
+			p->Members[i].NotLeader = (i != 0);
+			safestrncpy(p->Members[i].Name, cd.usersd[i]->status.name, NAME_LENGTH);
+			p->packetSize += sizeof(PACKET_ZC_ENTER_ROOM_sub);
 		}
 	}
-	WFIFOSET(fd, WFIFOW(fd, 2));
+	// Idk but client don't show my name without that and -1 count
```
Comment (AoShinRO) [inline src/map/clif.cpp:4587]: this needs to be studied

File: src/map/clif.cpp [hunk 2]
```diff
@@ -4557,38 +4557,39 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
 		return;
-	t = (int)(cd->owner->type == BL_NPC);
-	WFIFOHEAD(fd, 8 + (28*(cd->users+t)));
-	WFIFOW(fd, 0) = 0xdb;
-	WFIFOW(fd, 2) = 8 + (28*(cd->users+t));
-	WFIFOL(fd, 4) = cd->bl.id;
 
-	if(cd->owner->type == BL_NPC){
-		WFIFOL(fd, 30) = 1;
-		WFIFOL(fd, 8) = 0;
```
Comment (AoShinRO) [inline src/map/clif.cpp:4579]: anyone know what do this? 
WFIFOL(fd, 8) = 0;
commented header don't say about

File: src/map/clif.cpp [hunk 3]
```diff
@@ -4557,38 +4557,39 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
+	if (!session_isActive(sd.fd))
 		return;
-	t = (int)(cd->owner->type == BL_NPC);
-	WFIFOHEAD(fd, 8 + (28*(cd->users+t)));
-	WFIFOW(fd, 0) = 0xdb;
-	WFIFOW(fd, 2) = 8 + (28*(cd->users+t));
-	WFIFOL(fd, 4) = cd->bl.id;
 
-	if(cd->owner->type == BL_NPC){
-		WFIFOL(fd, 30) = 1;
```
Comment (AoShinRO) [inline src/map/clif.cpp:4578]: anyone know what do this?
WFIFOL(fd, 30) = 1;
commented header don't say about

File: src/map/clif.cpp (hunk 1 above)
Comment (AoShinRO) [reply src/map/clif.cpp]: solved adjusting packet size

File: src/map/clif.cpp [hunk 4]
```diff
@@ -4557,38 +4557,34 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
-			safestrncpy(WFIFOCP(fd, 8+(i+t)*28+4), cd->usersd[i]->status.name, NAME_LENGTH);
+	PACKET_ZC_ENTER_ROOM* p = reinterpret_cast<PACKET_ZC_ENTER_ROOM*>( packet_buffer );
+
+	int is_npc = (cd.owner->type == BL_NPC);
+	p->packetType = HEADER_ZC_ENTER_ROOM;
+	p->packetSize = sizeof(p) + (sizeof(PACKET_ZC_ENTER_ROOM_sub) * (cd.users + is_npc)) + sizeof(PACKET_ZC_ENTER_ROOM_sub);
+	p->chatId = cd.bl.id;
+	
+	if(is_npc){
+		p->Members[0].Flag = 0;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4573]:
```suggestion
		owner.flag = 0;
```

File: src/map/clif.cpp [hunk 5]
```diff
@@ -4557,38 +4557,34 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
+		safestrncpy(p->Members[0].Name, ((struct npc_data *)cd.owner)->name, NAME_LENGTH);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4574]:
```suggestion
		safestrncpy(owner.name, reinterpret_cast<npc_data*>(cd.owner)->name, sizeof(owner.name));
		p->packetSize += static_cast<decltype(p->packetSize)>( sizeof( owner ) );

```

File: src/map/clif.cpp [hunk 6]
```diff
@@ -4557,38 +4557,34 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
+		for (int i = 0; i < cd.users; i++) {
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4575]:
```suggestion
		for (size_t i = 0; i < cd.users; i++) {
			PACKET_ZC_ENTER_ROOM_sub& member = p->members[i + 1];

```

File: src/map/clif.cpp [hunk 7]
```diff
@@ -4557,38 +4557,34 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
+			p->Members[i+is_npc].Flag = 1;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4576]:
```suggestion
			member.flag = 1;
```

File: src/map/clif.cpp [hunk 8]
```diff
@@ -4557,38 +4557,34 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
+			safestrncpy(p->Members[i+is_npc].Name, cd.usersd[i]->status.name, NAME_LENGTH);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4577]:
```suggestion
			safestrncpy(member.name, cd.usersd[i]->status.name, sizeof(member.name));
```

File: src/map/clif.cpp [hunk 9]
```diff
@@ -4557,38 +4557,34 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
 		}
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4578]:
```suggestion

			p->packetSize += static_cast<decltype(p->packetSize)>( sizeof( member ) );
		}
```

File: src/map/clif.cpp [hunk 10]
```diff
@@ -4557,38 +4557,34 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
 	}else{
-		for (i = 0; i < cd->users; i++) {
-			WFIFOL(fd, 8+i*28) = (i != 0 || cd->owner->type == BL_NPC);
-			safestrncpy(WFIFOCP(fd, 8+(i+t)*28+4), cd->usersd[i]->status.name, NAME_LENGTH);
+		for (int i = 0; i < cd.users; i++) {
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4580]:
```suggestion
		for (size_t i = 0; i < cd.users; i++) {
			PACKET_ZC_ENTER_ROOM_sub& member = p->members[i];

```

File: src/map/clif.cpp [hunk 11]
```diff
@@ -4557,38 +4557,34 @@ void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result ){
...
+			p->Members[i].Flag = (i != 0);
+			safestrncpy(p->Members[i].Name, cd.usersd[i]->status.name, NAME_LENGTH);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:4582]:
```suggestion
			member.flag = i > 0;
			safestrncpy(member.name, cd.usersd[i]->status.name, sizeof(member.name));

			p->packetSize += static_cast<decltype(p->packetSize)>( sizeof( member ) );
```

File: src/map/chat.cpp [hunk 12]
```diff
@@ -168,7 +168,7 @@ int chat_joinchat(map_session_data* sd, int chatid, const char* pass)
 
 	pc_setchatid(sd,cd->bl.id);
 
-	clif_joinchatok(sd, cd); //To the person who newly joined the list of all
+	clif_joinchatok(*sd, *cd); //To the person who newly joined the list of all
```
Comment (Lemongrass3110) [inline src/map/chat.cpp:171]:
```suggestion
	// To the person who newly joined the chat
	clif_joinchatok(*sd, *cd);
```

File: src/map/clif.hpp [hunk 13]
```diff
@@ -884,7 +884,7 @@ void clif_GlobalMessage( block_list& bl, const char* message, enum send_target t
 void clif_createchat( map_session_data& sd, e_create_chatroom flag );
 void clif_dispchat(struct chat_data* cd, int fd);	// area or fd
 void clif_joinchatfail( map_session_data& sd, e_refuse_enter_room result );
-void clif_joinchatok(map_session_data *sd,struct chat_data* cd);	// self
+void clif_joinchatok(map_session_data& sd,chat_data& cd);	// self
```
Comment (Lemongrass3110) [inline src/map/clif.hpp:887]:
```suggestion
void clif_joinchatok(map_session_data& sd,chat_data& cd);
```

**Review by Lemongrass3110 — APPROVED** [review]

**Review by aleos89 — APPROVED** [review]

**Pull request merged**